asllib/TopoSort.ml

```diff
@@ -0,0 +1,212 @@
+module type OrderedHashedType = sig
+  include Set.OrderedType
+  include Hashtbl.HashedType with type t := t
+
+  val to_string : t -> string
+end
+
+module Make (O : OrderedHashedType) = struct
+  type succs = O.t -> O.t list
+
+  module OSet = Set.Make (O)
+  module OMap = Map.Make (O)
+  module OTbl = Hashtbl.Make (O)
+
+  module OStack = struct
+    type t = { stack : O.t Stack.t; mutable set : OSet.t }
+
+    let create () = { stack = Stack.create (); set = OSet.empty }
+
+    let push o t =
+      t.set <- OSet.add o t.set;
+      Stack.push o t.stack
+
+    let mem w t = OSet.mem w t.set
+
+    let pop t =
+      let o = Stack.pop t.stack in
+      t.set <- OSet.remove o t.set;
+      o
+
+    let pop_until =
+      let rec aux o acc t =
+        let o' = pop t in
+        let acc = o' :: acc in
+        if O.compare o' o = 0 then acc else aux o acc t
+      in
+      fun o t -> aux o [] t
+  end
+
+  type data = {
+    (* order from which nodes are discovered *)
+    index : int;
+    (* lowest index of nodes in the stack reachable from a node. *)
+    lowlink : int;
+  }
+  (** Book-keeping for each node *)
+
+  type state = { data : data OTbl.t; stack : OStack.t; index : int ref }
+  (** All book-keeping *)
+
+  let set_lowlink s v v_data lowlink =
+    let v_data = { v_data with lowlink } in
+    OTbl.replace s.data v v_data;
+    v_data
+
+  let rec each_successor succs s fold v (acc, v_data) w =
+    let () =
+      if false then
+        Format.eprintf "each_successor of v=%s w=%s@." (O.to_string v)
+          (O.to_string w)
+    in
+    match OTbl.find_opt s.data w with
+    | None ->
+        let acc, w_data = strong_connected succs s fold acc w in
+        if v_data.lowlink > w_data.lowlink then
+          (acc, set_lowlink s v v_data w_data.lowlink)
+        else (acc, v_data)
+    | Some w_data ->
+        if OStack.mem w s.stack && v_data.lowlink > w_data.index then
+          (acc, set_lowlink s v v_data w_data.index)
+        else (acc, v_data)
+
+  and strong_connected succs s fold acc v =
+    (* Assumption that [acc.datas] does not contain [v]. *)
```

Comment:
Shouldn't [acc.datas] be [s.data]? This would justify using the `add` function of hash tables below.